auth_mellon_diagnostics.c

```diff
@@ -0,0 +1,807 @@
+    case APLOG_TRACE2:  return "APLOG_TRACE2";
+    case APLOG_TRACE3:  return "APLOG_TRACE3";
+    case APLOG_TRACE4:  return "APLOG_TRACE4";
+    case APLOG_TRACE5:  return "APLOG_TRACE5";
+    case APLOG_TRACE6:  return "APLOG_TRACE6";
+    case APLOG_TRACE7:  return "APLOG_TRACE7";
+    case APLOG_TRACE8:  return "APLOG_TRACE8";
+    default:
+        return apr_psprintf(r->pool, "APLOG_%d", level);
+    }
+}
+
+static void
+am_diag_initilaize_req(request_rec *r, apr_file_t *diag_fd,
+                       am_req_cfg_rec *req_cfg);
+
+apr_size_t
+am_diag_time_t_to_8601_buf(char *buf, apr_size_t buf_size, apr_time_t t)
+{
+    apr_size_t ret_size;
+    apr_time_exp_t tm;
+    //const char fmt[] = "%Y-%m-%dT%H:%M:%SZ";
+    const char fmt[] = "%FT%TZ";
+
+    apr_time_exp_gmt(&tm, t);
+    apr_strftime(buf, &ret_size, buf_size, fmt, &tm);
+    if (ret_size == 0) buf[0] = 0; /* on errror assure string is null terminated */
+    return ret_size;
+}
+
+char *
+am_diag_time_t_to_8601(request_rec *r, apr_time_t t)
+{
+    char *buf;
+
+    buf = apr_palloc(r->pool, ISO_8601_BUF_SIZE);
+    if (!buf) return NULL;
+
+    am_diag_time_t_to_8601_buf(buf, ISO_8601_BUF_SIZE, t);
+    return buf;
+}
+
+static int
+am_diag_open_log(server_rec *s, apr_pool_t *p)
+{
+
+    am_mod_cfg_rec *mod_cfg = am_get_mod_cfg(s);
+
+    if (!AM_DIAG_ENABLED(mod_cfg)) {
+        ap_log_error(APLOG_MARK, APLOG_DEBUG, 0, s,
+                     "mellon diagnostics disabled for server %s", s->defn_name);
+        return 1;
+    } else {
+        ap_log_error(APLOG_MARK, APLOG_DEBUG, 0, s,
+                     "mellon diagnostics enabled for server %s"
+                     " diag_filename=%s", s->defn_name, mod_cfg->diag_filename);
+    }
+
+    if (!mod_cfg->diag_filename || mod_cfg->diag_fd)
+        return 1;
+
+    if (*mod_cfg->diag_filename == '|') {
+        piped_log *pl;
+        const char *pname = ap_server_root_relative(p, mod_cfg->diag_filename + 1);
+
+        pl = ap_open_piped_log(p, pname);
+        if (pl == NULL) {
+            ap_log_error(APLOG_MARK, APLOG_ERR, 0, s,
+                         "couldn't spawn mellon diagnostics log pipe %s",
+                         mod_cfg->diag_filename);
+            return 0;
+        }
+        mod_cfg->diag_fd = ap_piped_log_write_fd(pl);
+    }
+    else {
+        const char *fname = ap_server_root_relative(p, mod_cfg->diag_filename);
+        apr_status_t rv;
+
+        if ((rv = apr_file_open(&mod_cfg->diag_fd, fname,
+                                APR_WRITE | APR_APPEND | APR_CREATE,
+                                APR_OS_DEFAULT, p)) != APR_SUCCESS) {
+            ap_log_error(APLOG_MARK, APLOG_ERR, rv, s,
+                         "could not open mellon diagnostics log file %s.",
+                         fname);
+            return 0;
+        }
+    }
+
+    return 1;
+}
+
+
+int
+am_diag_log_init(apr_pool_t *pc, apr_pool_t *p, apr_pool_t *pt, server_rec *s)
+{
+    for ( ; s ; s = s->next) {
```

Comment:
You were right to be confused because I was confused between module specific and server specific configuration. I was using mod_log_forensic as a model. I have reworked this portion of the code. The intent is to have the logging be server specific, this way you can control logging based on whether you're in the core server or in a specific virtual server, being able to enable logging and select the log file based on server configuration is valuable. To have the logging be server specific I moved the logging members out of am_mod_cfg_rec and into am_srv_cfg_rec and also coalesced the logging members into their own struct. This also required implementing a server config merge function and adding it to the module callbacks. I have verified with the code changes one can control the logging flags and file from within a server configuration block. I also improved the error log message to indicate which server is being initialized, where the configuration directives are coming from and what the hostname and port are for the server.